src/napari/_vispy/canvas.py

```diff
@@ -875,41 +903,109 @@ def _update_layer_overlays(self, layer: Layer) -> None:
-                layer, {}
-            ).items():
+        # then gridded viewer overlays and layer overlays, by viewbox, in order
+        for viewbox_idx, (_, layer_indices) in enumerate(
+            self.viewer.grid.iter_viewboxes(len(self.viewer.layers))
+        ):
+            if not layer_indices:
+                # last empty boxes of the grid
+                break
+            view = viewbox_idx if self.viewer.grid.enabled else None
+
+            for overlay, vispy_overlays in self._overlay_to_visual.items():
                 if (
-                    layer.visible
-                    and overlay.visible
+                    overlay.visible
                     and isinstance(overlay, CanvasOverlay)
                     and overlay.position in list(CanvasPosition)
+                    and (overlay.gridded and self.viewer.grid.enabled)
                 ):
-                    yield overlay, [vispy_overlay]
+                    yield overlay, vispy_overlays[viewbox_idx], view
+
+            for layer_idx in layer_indices:
+                layer = self.viewer.layers[layer_idx]
+                for (
+                    overlay,
+                    vispy_overlay,
+                ) in self._layer_overlay_to_visual.get(layer, {}).items():
+                    if (
+                        layer.visible
+                        and overlay.visible
+                        and isinstance(overlay, CanvasOverlay)
+                        and overlay.position in list(CanvasPosition)
+                    ):
+                        yield overlay, vispy_overlay, view
 
     def _update_overlay_canvas_positions(self, event=None):
+        # TODO: make settable
+        x_padding = y_padding = 10.0
+        x_offset_total = {}
+        y_offset_total = {}
         for (
-            _,
-            vispy_overlays,
+            overlay,
+            vispy_overlay,
+            view,
         ) in self._get_ordered_visible_canvas_overlays():
-            for vispy_overlay in vispy_overlays:
-                vispy_overlay._on_position_change()
+            # TODO: vertical vs horizontal tiling should be settable!
+            x_offset_total.setdefault(
+                view, dict.fromkeys(CanvasPosition, x_padding)
+            )
+            y_offset_total.setdefault(
+                view, dict.fromkeys(CanvasPosition, y_padding)
+            )
+
+            x_offset = x_offset_total[view][overlay.position]
+            y_offset = y_offset_total[view][overlay.position]
+
+            # add offset to the following overlays based on tiling direction
+            if overlay.position in ('top_right', 'bottom_left'):
+                x_offset_total[view][overlay.position] += (
+                    vispy_overlay.x_size + x_padding
+                )
+            else:
+                y_offset_total[view][overlay.position] += (
+                    vispy_overlay.y_size + y_padding
+                )
+
```

Comment:
I still do not understand why these cases are needed.

Reply:
right now tiling direction is hardcoded based on the corner (there's a todo a few lines above for this). The `top_right` and `bottom_left` corner will tile horizontally, while the others vertically. This will be replaced by something like the following:

```py
if <tiling_settings>[overlay.position] == 'horizontal':
	...
```